From 409fbe390833771c39ca36cfc4977a9274dcfa33 Mon Sep 17 00:00:00 2001
From: Karthikeyan K <karthikk@habana.ai>
Date: Mon, 07 Feb 2022 07:29:11 +0200
Subject: [PATCH] [SW-68803] boot_desc: move init to common init

Moving the boot desc initialization function of spi fw image to the
common arboot rom init as this needs to be only done once. An
indication api is added to check before reading the boot desc data.

Change-Id: I93f9d70e3fb8325babaf2b71e33316c5b8a7a688
---

diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
index 4f8289d..a3018f2 100644
--- a/arch/gaudi3/arc/common.c
+++ b/arch/gaudi3/arc/common.c
@@ -163,6 +163,9 @@
 	arcboot_rom_init_sp_sram();
 
 	spi_flash_init();
+
+	if (!boot_desc_init(HL_FW_BOOT_PTR_SPI))
+		arcboot_rom_status_update(PROGRESS_INIT_SPI_BOOT_DESC_FAIL);
 }
 
 void arcboot_rom_fuse(void)
@@ -319,6 +322,14 @@
 	/* TODO: Add tpm relavent code */
 }
 
+static int arcboot_rom_boot_desc_init_done(enum hl_fw_boot_ptr_imgs img)
+{
+	struct hl_fw_boot_ptrs_ctx *p_img_ptr_ctx =
+			&arcboot_rom_global.boot_ptr_ctx[img];
+
+	return !!p_img_ptr_ctx->boot_desc_flag;
+}
+
 void arcboot_rom_bringup_pcie(void)
 {
 	void (*ppboot)();
@@ -326,8 +337,7 @@
 			&arcboot_rom_global.boot_ptr_ctx[HL_FW_BOOT_PTR_SPI].load_addr;
 	int ret;
 
-	if (!boot_desc_init(HL_FW_BOOT_PTR_SPI)) {
-		arcboot_rom_status_update(PROGRESS_BRINGUP_PCI_BOOT_DESC_FAIL);
+	if (arcboot_rom_boot_desc_init_done(HL_FW_BOOT_PTR_SPI)) {
 		goto skip_bringup;
 	}
 
diff --git a/include/arcboot.h b/include/arcboot.h
index 2543435..578ca42 100644
--- a/include/arcboot.h
+++ b/include/arcboot.h
@@ -90,18 +90,18 @@
 	PROGRESS_INIT = 0,
 	PROGRESS_INIT_SP_SRAM_NOT_COMPLETED = 1,
 	PROGRESS_INIT_SPI_BOOT_DESC_DMA_START = 2,
-	PROGRESS_INIT_SPI_BOOT_DESC_DMA_FINISH = 3,
-	PROGRESS_INIT_EFUSE_SPI = 4,
-	PROGRESS_REVOKE_LOC = 5,
-	PROGRESS_REVOKE_LOC_0 = 6,
-	PROGRESS_REVOKE_LOC_1 = 7,
-	PROGRESS_KEY_LOC = 8,
-	PROGRESS_KEY_LOC_0 = 9,
-	PROGRESS_KEY_LOC_1 = 10,
-	PROGRESS_VER_LOC = 11,
-	PROGRESS_VER_LOC_0 = 12,
-	PROGRESS_VER_LOC_1 = 13,
-	PROGRESS_BRINGUP_PCI_BOOT_DESC_FAIL = 14,
+	PROGRESS_INIT_SPI_BOOT_DESC_FAIL = 3,
+	PROGRESS_INIT_SPI_BOOT_DESC_DMA_FINISH = 4,
+	PROGRESS_INIT_EFUSE_SPI = 5,
+	PROGRESS_REVOKE_LOC = 6,
+	PROGRESS_REVOKE_LOC_0 = 7,
+	PROGRESS_REVOKE_LOC_1 = 8,
+	PROGRESS_KEY_LOC = 9,
+	PROGRESS_KEY_LOC_0 = 10,
+	PROGRESS_KEY_LOC_1 = 11,
+	PROGRESS_VER_LOC = 12,
+	PROGRESS_VER_LOC_0 = 13,
+	PROGRESS_VER_LOC_1 = 14,
 	PROGRESS_BRINGUP_PCI_PRI = 15,
 	PROGRESS_BRINGUP_PCI_PRI_TOO_BIG = 16,
 	PROGRESS_BRINGUP_PCI_PRI_DMA = 17,
